In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from zipfile import ZipFile
with ZipFile('/content/drive/MyDrive/Project files/archive.zip','r') as zipObj:
  zipObj.extractall('/content/drive/MyDrive/Project files/Bankproject')

In [2]:
df=pd.read_excel('/content/drive/MyDrive/Project files/Bankproject/bank.xlsx')
df.head()

Account No       DATE               TRANSACTION DETAILS  CHQ.NO.  \
0  409000611074' 2017-06-29  TRF FROM  Indiaforensic SERVICES      NaN   
1  409000611074' 2017-07-05  TRF FROM  Indiaforensic SERVICES      NaN   
2  409000611074' 2017-07-18        FDRL/INTERNAL FUND TRANSFE      NaN   
3  409000611074' 2017-08-01   TRF FRM  Indiaforensic SERVICES      NaN   
4  409000611074' 2017-08-16        FDRL/INTERNAL FUND TRANSFE      NaN   

  VALUE DATE  WITHDRAWAL AMT  DEPOSIT AMT  BALANCE AMT  .  
0 2017-06-29             NaN    1000000.0    1000000.0  .  
1 2017-07-05             NaN    1000000.0    2000000.0  .  
2 2017-07-18             NaN     500000.0    2500000.0  .  
3 2017-08-01             NaN    3000000.0    5500000.0  .  
4 2017-08-16             NaN     500000.0    6000000.0  .

In [3]:
#Finding the shape of the dataset
df.shape

(116201, 9)

In [4]:
df.describe()

CHQ.NO.  WITHDRAWAL AMT   DEPOSIT AMT   BALANCE AMT
count     905.000000    5.354900e+04  6.265200e+04  1.162010e+05
mean   791614.503867    4.489190e+06  3.806586e+06 -1.404852e+09
std    151205.932910    1.084850e+07  8.683093e+06  5.348202e+08
min         1.000000    1.000000e-02  1.000000e-02 -2.045201e+09
25%    704231.000000    3.000000e+03  9.900000e+04 -1.690383e+09
50%    873812.000000    4.708300e+04  4.265000e+05 -1.661395e+09
75%    874167.000000    5.000000e+06  4.746411e+06 -1.236888e+09
max    874525.000000    4.594475e+08  5.448000e+08  8.500000e+06

In [5]:
#Searching for null values
df.isnull().sum()

Account No                  0
DATE                        0
TRANSACTION DETAILS      2499
CHQ.NO.                115296
VALUE DATE                  0
WITHDRAWAL AMT          62652
DEPOSIT AMT             53549
BALANCE AMT                 0
.                           0
dtype: int64

Because of 115296 null values in CHQ.NO.out of 116201 rows,we are going to drop this column.

In [6]:
#Dropping CHQ.NO. column
df.drop('CHQ.NO.',axis=1,inplace=True)

In [7]:
#Now,we are going to the fill the null values with median of the respective columns
df['DEPOSIT AMT'].fillna(df['DEPOSIT AMT'].median(),inplace=True)
df['WITHDRAWAL AMT'].fillna(df['WITHDRAWAL AMT'].median(),inplace=True)

In [8]:
df.isnull().sum()

Account No                0
DATE                      0
TRANSACTION DETAILS    2499
VALUE DATE                0
WITHDRAWAL AMT            0
DEPOSIT AMT               0
BALANCE AMT               0
.                         0
dtype: int64

**As transaction details is a string column no need to worry about the null values**

In [9]:
#Let's set Account No as index for better understanding
df.set_index('Account No',inplace=True)

In [10]:
df.head()

DATE               TRANSACTION DETAILS VALUE DATE  \
Account No                                                              
409000611074' 2017-06-29  TRF FROM  Indiaforensic SERVICES 2017-06-29   
409000611074' 2017-07-05  TRF FROM  Indiaforensic SERVICES 2017-07-05   
409000611074' 2017-07-18        FDRL/INTERNAL FUND TRANSFE 2017-07-18   
409000611074' 2017-08-01   TRF FRM  Indiaforensic SERVICES 2017-08-01   
409000611074' 2017-08-16        FDRL/INTERNAL FUND TRANSFE 2017-08-16   

               WITHDRAWAL AMT  DEPOSIT AMT  BALANCE AMT  .  
Account No                                                  
409000611074'         47083.0    1000000.0    1000000.0  .  
409000611074'         47083.0    1000000.0    2000000.0  .  
409000611074'         47083.0     500000.0    2500000.0  .  
409000611074'         47083.0    3000000.0    5500000.0  .  
409000611074'         47083.0     500000.0    6000000.0  .

In [ ]:
df.drop_duplicates()

DATE               TRANSACTION DETAILS VALUE DATE  \
Account No                                                              
409000611074' 2017-06-29  TRF FROM  Indiaforensic SERVICES 2017-06-29   
409000611074' 2017-07-05  TRF FROM  Indiaforensic SERVICES 2017-07-05   
409000611074' 2017-07-18        FDRL/INTERNAL FUND TRANSFE 2017-07-18   
409000611074' 2017-08-01   TRF FRM  Indiaforensic SERVICES 2017-08-01   
409000611074' 2017-08-16        FDRL/INTERNAL FUND TRANSFE 2017-08-16   
...                  ...                               ...        ...   
409000362497' 2019-03-05  TRF TO 1196428  Indiaforensic SE 2019-03-05   
409000362497' 2019-03-05        FDRL/INTERNAL FUND TRANSFE 2019-03-05   
409000362497' 2019-03-05        FDRL/INTERNAL FUND TRANSFE 2019-03-05   
409000362497' 2019-03-05                 IMPS 05-03-20194C 2019-03-05   
409000362497' 2019-03-05         Sweep Trf To: 40900036427 2019-03-05   

               WITHDRAWAL AMT  DEPOSIT AMT   BALANCE AMT  .  
Account No                                                   
409000611074'        47083.00    1000000.0  1.000000e+06  .  
409000611074'        47083.00    1000000.0  2.000000e+06  .  
409000611074'        47083.00     500000.0  2.500000e+06  .  
409000611074'        47083.00    3000000.0  5.500000e+06  .  
409000611074'        47083.00     500000.0  6.000000e+06  .  
...                       ...          ...           ... ..  
409000362497'       117934.30     426500.0 -1.901902e+09  .  
409000362497'        47083.00     300000.0 -1.901602e+09  .  
409000362497'        47083.00     300000.0 -1.901302e+09  .  
409000362497'       109868.65     426500.0 -1.901412e+09  .  
409000362497'         5000.00     426500.0 -1.901417e+09  .  

[116162 rows x 7 columns]

In [29]:
print(df['DATE'].min())
print(df['DATE'].max())

2015-01-01 00:00:00
2019-03-05 00:00:00


**We are provided with 4 years transaction history**

In [21]:
df['WITHDRAWAL AMT'].mean()

2094142.704458299

In [28]:
#Let's find out withdrawl amount greater than its mean.
withdraw=df[df['WITHDRAWAL AMT']>df['WITHDRAWAL AMT'].mean()*2]
withdraw

DATE               TRANSACTION DETAILS VALUE DATE  \
Account No                                                              
409000425051' 2018-10-26             TRF TO Rajendra Kumar 2018-10-26   
409000425051' 2018-10-31                TRF TO  Myur Joshi 2018-10-31   
409000405747' 2016-01-30                 TRF TO Maur Joshi 2016-01-30   
409000405747' 2016-01-30  TRF TO Indiaforensic SERVICES IN 2016-01-30   
409000405747' 2018-09-24  FUNDS TRF  Indiaforensic SERVICE 2018-09-24   
...                  ...                               ...        ...   
409000362497' 2018-08-14                     IMPS 1408183C 2018-08-14   
409000362497' 2018-08-16                     IMPS 1608183C 2018-08-16   
409000362497' 2018-08-20                     IMPS 2008183C 2018-08-20   
409000362497' 2019-01-08  TRF TO 1196428  Indiaforensic SE 2019-01-08   
409000362497' 2019-01-09  Indiaforensic SERVICES INDIA PVT 2019-01-09   

               WITHDRAWAL AMT  DEPOSIT AMT   BALANCE AMT  .  
Account No                                                   
409000425051'    1.500000e+07     426500.0 -1.454080e+07  .  
409000425051'    3.540000e+08     426500.0 -3.535408e+08  .  
409000405747'    1.700000e+08     426500.0 -5.267348e+08  .  
409000405747'    3.000000e+07     426500.0 -5.567348e+08  .  
409000405747'    1.000000e+07     426500.0 -3.667145e+08  .  
...                       ...          ...           ... ..  
409000362497'    4.189678e+06     426500.0 -1.904609e+09  .  
409000362497'    4.751486e+06     426500.0 -1.902778e+09  .  
409000362497'    4.872862e+06     426500.0 -1.902095e+09  .  
409000362497'    8.197253e+06     426500.0 -1.898754e+09  .  
409000362497'    5.833402e+06     426500.0 -1.901558e+09  .  

[14217 rows x 7 columns]

**In my point of view,I extract the above rows from dataset whose withdrawl amount are double the mean.If any fraud transaction occurs in the dataset happened in this 14k rows only,the reamining are artlessness transactions.**

In [12]:
pivot1=pd.pivot_table(df,index='WITHDRAWAL AMT')
pivot1

BALANCE AMT  DEPOSIT AMT
WITHDRAWAL AMT                           
1.000000e-02   -5.452654e+08     426500.0
2.000000e-02   -5.428633e+08     426500.0
3.000000e-02   -5.458076e+08     426500.0
4.000000e-02   -5.432870e+08     426500.0
2.000000e-01   -5.479048e+08     426500.0
...                      ...          ...
2.671403e+08   -9.021087e+08     426500.0
3.540000e+08   -3.535408e+08     426500.0
4.000000e+08   -5.084147e+08     426500.0
4.482072e+08   -2.036201e+09     426500.0
4.594475e+08   -1.929281e+09     426500.0

[20881 rows x 2 columns]

In [20]:
#Let's create pivot table for withdrawl and balance amount
pivot2=pd.pivot_table(df,index='WITHDRAWAL AMT',values='BALANCE AMT')
pivot2

BALANCE AMT
WITHDRAWAL AMT              
1.000000e-02   -5.452654e+08
2.000000e-02   -5.428633e+08
3.000000e-02   -5.458076e+08
4.000000e-02   -5.432870e+08
2.000000e-01   -5.479048e+08
...                      ...
2.671403e+08   -9.021087e+08
3.540000e+08   -3.535408e+08
4.000000e+08   -5.084147e+08
4.482072e+08   -2.036201e+09
4.594475e+08   -1.929281e+09

[20881 rows x 1 columns]

In [18]:
print('maximum.   ',pivot2.max())
print('minimum.   ',pivot2.min())
print('median.   ',pivot2.median())
print('mean.      ',pivot2.mean())

maximum.    BALANCE AMT    3273973.03
dtype: float64
minimum.    BALANCE AMT   -2.045181e+09
dtype: float64
median.    BALANCE AMT   -1.584995e+09
dtype: float64
mean.       BALANCE AMT   -1.224141e+09
dtype: float64


**The mean value of balance amount in all accounts is negative value it shows that the bank might be at risk.**

In [31]:
pivot3=pd.pivot_table(df,index='WITHDRAWAL AMT',values='BALANCE AMT',aggfunc=['sum','mean'])
pivot3

sum          mean
                 BALANCE AMT   BALANCE AMT
WITHDRAWAL AMT                            
1.000000e-02   -1.635796e+09 -5.452654e+08
2.000000e-02   -5.428633e+08 -5.428633e+08
3.000000e-02   -5.458076e+08 -5.458076e+08
4.000000e-02   -1.086574e+09 -5.432870e+08
2.000000e-01   -5.479048e+08 -5.479048e+08
...                      ...           ...
2.671403e+08   -9.021087e+08 -9.021087e+08
3.540000e+08   -3.535408e+08 -3.535408e+08
4.000000e+08   -5.084147e+08 -5.084147e+08
4.482072e+08   -2.036201e+09 -2.036201e+09
4.594475e+08   -1.929281e+09 -1.929281e+09

[20881 rows x 2 columns]